src/altrep-sparse-real.c

```diff
@@ -0,0 +1,176 @@
+#define R_NO_REMAP
+#include "R.h"
+#include <Rinternals.h>
+#include <R_ext/Altrep.h>
+
+// Initialised at load time
+R_altrep_class_t altrep_sparse_real_class;
+
+SEXP ffi_altrep_new_sparse_real(SEXP x) {
+  return R_new_altrep(altrep_sparse_real_class, x, R_NilValue);
+}
+
+SEXP alrep_sparse_real_Materialize(SEXP vec) {
+  SEXP out = R_altrep_data2(vec);
+
+  if (out != R_NilValue) {
+    return out;
+  }
+
+  SEXP data1 = R_altrep_data1(vec);
+  SEXP val = VECTOR_ELT(data1, 0);
+  SEXP pos = VECTOR_ELT(data1, 1);
+  SEXP len = VECTOR_ELT(data1, 2);
+
+  out = PROTECT(Rf_allocVector(REALSXP, Rf_asInteger(len)));
+  
+  int n = Rf_asInteger(len);
+  
+  for (int i = 0; i < n; ++i) {
+    SET_REAL_ELT(out, i, 0);
+  }
+
+  int m = Rf_length(pos);
+
+  for (int i = 0; i < m; ++i) {
+    SET_REAL_ELT(out, INTEGER_ELT(pos, i) - 1, REAL_ELT(val, i));
+  }
+
+  R_set_altrep_data2(vec, out);
+
+  UNPROTECT(1);
+  return out;
+}
+
+// -----------------------------------------------------------------------------
+// ALTVEC
+
+void* altrep_sparse_real_Dataptr(SEXP x, Rboolean writeable) {
+  return STDVEC_DATAPTR(alrep_sparse_real_Materialize(x));
+}
+
+const void* altrep_sparse_real_Dataptr_or_null(SEXP vec) {
+    SEXP out = R_altrep_data2(vec);
+
+  if (out == R_NilValue) {
+    return NULL;
+  } else {
+    return out;
+  }
+}
+
+static SEXP altrep_sparse_real_Extract_subset(SEXP x, SEXP indx, SEXP call) {
+
+  SEXP data1 = R_altrep_data1(x);
+  SEXP val_old = VECTOR_ELT(data1, 0);
+  SEXP pos_old = VECTOR_ELT(data1, 1);
+  SEXP matches = Rf_match(pos_old, indx, R_NaInt);
```

Comment:
I think you can also use the binary search approach here, and it would be super nice. What I imagine is a loop like

```
for (i in seq_along(indx)) {
  match[i] = binary_search(indx[i], pos)
}
```

Where `binary_search()` is a C function that returns:
- `NA_INTEGER` if it can't find the value
- The C indexed location in `pos` of the value

You could reuse that same `binary_search()` function in the Elt method

Reply:
You'd also be able to count up the number of matches in that loop directly rather than in the `n` loop below, like

```
match[i] = binary_search(indx[i], pos)
count = count + (match[i] != NA_INTEGER)
```

Reply:
Oh I'm not sure about the guarantees on `Extract_subset` but in theory `indx` may contain `NA_INTEGER` or an index that is greater than the length of your vector.

i.e. `(1:3)[4]` or `(1:3)[NA_integer_]`

and I'm pretty sure you don't account for that

For example, both of those would be "no match" cases but you _do_ want to allocate space for them in the result (both result in NA values)

Reply:
Then the binary search loop could be something like

```
for (i in seq_along(indx)) {
  if (indx[i] > length) {
    match[i] = NA;
    continue;
  }
  if (index[i] == NA) {
    match[i] = NA;
    continue;
  }
  match[i] = binary_search(indx[i], pos)
}
```

Where this time `binary_search()` would return `length(pos)` in the no match case (that's how a typical binary search works) allowing you to differentiate between the 3 cases of:
- Index was OOB or NA (`NA` value in result)
- Index was located in `pos` (`val` value in result)
- Index was `< len`, so could have had a value in `pos`, but does not appear in `pos`, so we ignore it

Reply:
If handling `NA` and OOB values seems too complicated to mix in, then one option is to do an initial loop over `indx` and look for `NA` or OOB indices and if you detect any then you return `NULL` from the function immediately, which is the ALTREP way of saying "i give up, handle this with the fallback method" which will materialize the vector. That's totally reasonable and is what vroom does.

Then you'd just handle values in the range of `1 -> len`